src/interactive/azext_interactive/azclishell/recommendation.py

```diff
@@ -0,0 +1,392 @@
+            processed_exception = self.cur_thread.processed_exception
+            if not recommendations or not scenario:
+                send_feedback(-1, trigger_commands, processed_exception, recommendations, rec=None)
+            else:
+                send_feedback(f'b{scenario_idx+1}', trigger_commands, processed_exception, recommendations, rec=scenario)
+
+    def update_executing(self, cmd, feedback=True):
+        """Update executing command info, and prefetch the recommendation result as if the execution is successful
+
+        :param cmd: The command that would be executed soon
+        :param feedback: whether send feedback to telemetry using this command as actual execution vs
+        previous recommendation
+        """
+        if not self.enabled:
+            return
+        command = re.sub('^az *', '', cmd).split('-', 1)[0].strip()
+        param = sorted([p for p in cmd.split() if p.startswith('-')])
+        if feedback:
+            self._feedback_command(command)
+        self.executing_command = {'command': command, 'arguments': param}
+        self._update()
+
+    def update_exec_result(self, exit_code, result_summary=''):
+        """Update execution result of executing_command set previously, fetch recommendation result if execution fails
+
+        :param exit_code: exit code of previous executing command
+        :param result_summary: Result Summary of previous executing command
+        """
+        if not self.enabled or not self.executing_command:
+            return
+        self.rec_path.append_result(self.executing_command['command'], self.executing_command['arguments'],
+                                    exit_code, result_summary)
+        self.executing_command = None
+        if exit_code != 0:
+            self._update()
+
+    def _update(self):
+        """Update recommendation result in new thread"""
+        self.cur_thread = RecommendThread(self.cli_ctx, self.rec_path, self.executing_command,
+                                          self.on_recommendation_prepared)
+        self.cur_thread.start()
+
+    def _get_result(self, non_block=True, timeout=3.0, rec_type=RecommendType.Command):
+        if not self.cur_thread:
+            if non_block:
+                return None
+            else:
+                self._update()
+        if not non_block:
+            self.cur_thread.join(timeout)
+        if not self.cur_thread.result:
+            return self.cur_thread.result
+        return [rec for rec in self.cur_thread.result if rec['type'] == rec_type]
+
+    def get_commands(self, non_block=True, timeout=3.0):
+        """
+        Get the latest recommended commands
+        :param non_block: whether to wait for data to be prepared
+        :param timeout: block timeout
+        :return: recommendation or None if the result is not prepared
+        """
+        if not self.enabled:
+            return []
```

Comment:
>     def _get_result(self, non_block=True, timeout=3.0, rec_type=RecommendType.Command):
        if not self.cur_thread:
            if non_block:
                return None

May I ask when do we need to return `None` or `[]`? Could their return values be unified?

Reply:
For `_get_result`, `[]` is different from `None`. `[]` represents no recommendation fetched from remote while `None` represents the request is running and recommendation is not ready.

Reply:
OK, please add some comments to explain the difference between them